tests/unit/test_ncp_srp_server.cpp

```diff
@@ -0,0 +1,132 @@
+/*
+ *  Copyright (c) 2024, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <stdio.h>
+
+#include <openthread/srp_server.h>
+
+#include "test_platform.h"
+#include "test_util.h"
+#include "common/code_utils.hpp"
+#include "lib/spinel/spinel_buffer.hpp"
+#include "lib/spinel/spinel_encoder.hpp"
+#include "ncp/ncp_base.hpp"
+
+#if OPENTHREAD_CONFIG_SRP_SERVER_ENABLE
+
+namespace ot {
+
+constexpr uint16_t kMaxSpinelBufferSize = 2048;
+
+static otError GenerateSpinelSrpServerFrame(spinel_prop_key_t aProp, bool aEnable, uint8_t *aBuf, uint16_t &aLen)
```

Comment:
Should we add an `#if` check guard so that this whole test is skipped if NCP (spinel) is not enabled/included in a build? I think we can have such a configs where we build CLI only (along with all unit tests?)

Reply:
I noticed we have `ot_unit_ncp_test()` in `CMakeLists.txt`. I guess it does limit the unit test to when "ncp" lib is built. Is this correct?

Reply:
Good point! I just checked and I found that currently the cli library `openthread-cli-{ftd,mtd}` and ncp libray`openthread-ncp-{ftd,mtd}` will be always built and not controlled by any cmake conditions.

The cmake option `OT_APP_CLI` and `OT_APP_NCP` only controls the build of the app executable. But the cli library and ncp library will be built anyway.

In `src/CMakeLists.txt`, it unconditionally adds the ncp directory
```
add_subdirectory(ncp)
```
And in `src/ncp/CMakeLists`, it unconditionally builds ncp-ftd or ncp-mtd library based on `OT_FTD` and `OT_MTD`:
```
if(OT_FTD)
    include(ftd.cmake)
endif()

if(OT_MTD)
    include(mtd.cmake)
endif()
```

So the "ncp" lib is always built... I think we can do an enhancement to only build the ncp lib when NCP app is to be built. And we use the same condition to control these ncp unit test. Maybe in another PR. Thoughts?